### PR TITLE
feat(server): single-tenant REST API server over the _app/ layer (ADR-0021)

### PR DESCRIPTION
**Pull request by teng-lin**
## Summary

Adds a **single-tenant REST API server** as the third sibling adapter over the transport-neutral `_app/` layer (after `cli/` and the drafted MCP server) — the adapter ADR-0021 anticipated. It's a thin FastAPI app: routes are adapters over the existing `_app` cores + public client namespaces, so no business logic is duplicated.

Ships behind an optional `server` extra (`pip install notebooklm-py[server]`) and a `notebooklm-server` console script. Nothing changes for existing CLI/library users when the extra is absent.

## Surface (`/v1`, bearer-gated)

- **notebooks** — list / get / create / delete
- **sources** — list / get / add (url · text · file) / delete
- **chat** — blocking `ask` (no SSE in v1)
- **artifacts** — list / generate (non-blocking → `task_id`) / poll / download
- **`/healthz`** — public liveness probe (outside `/v1`, no token, no account info)

## Key design

- **One client per process, bound at lifespan.** A single `NotebookLMClient.from_storage()` is opened inside the server loop (ADR-0004 loop-affinity) and reused for the process lifetime.
- **Poll-the-resource async model** (no `/jobs`). An in-process pending-id registry resolves the 200-pending-vs-404-never-created ambiguity that `get_or_none → None` alone cannot. Registry is FIFO-bounded.
- **Auth:** required static bearer token (constant-time compare) + loopback-Host DNS-rebinding guard; fails closed when no token is configured. Config via `NOTEBOOKLM_SERVER_*` env vars.
- **Errors project from `_app.errors.classify`** onto HTTP status + a uniform `{"error": {"category", "message"}}` envelope — the category is classified once, never re-derived. Includes the shared fix mapping a normalized gRPC-status-5 `ClientError` → `NOT_FOUND` (was mis-mapped), which also benefits CLI/MCP.
- **Uploads** spool to `0o600` unique temp files under a size cap, with a Content-Length pre-check as the disk-exhaustion guard.
- **No unauthenticated schema surface** — `/docs`, `/redoc`, `/openapi.json` are disabled.

## Guardrails

- `_app` must not import fastapi (`test_app_boundary` + `test_server_boundary`).
- `test_server_classify_consistency` pins every `ErrorCategory` to a status.
- Module-size ratchet, architecture-map freshness, api-compat (new package auto-excluded) all honored.
- Every wire-bound error message is run through `scrub_secrets` (honors the #1517/#1518 redaction discipline); the token/Authorization header are never logged.

## Testing

New `tests/server/` suite (importorskip-guarded so the fastapi-free canonical install still collects cleanly), including `test_hardening.py` (host-guard edge cases, error scrubbing, envelope shape, pre-buffer 413, bounded registry) and `test_e2e_loop.py` (single-client loop affinity). Full suite: **10002 passed, 71 skipped, 1 xfailed**. ruff + mypy clean.

Built via ce-brainstorm → ce-plan → momus (codex) review → ce-work, then the full `/polish` pipeline (simplify → review → ultrathink).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a standalone REST server (notebooklm-server): token auth, loopback bind guard, /v1 APIs for notebooks/sources/chat/artifacts, health endpoint, upload limits, polling model, artifact download, pending-registry, and a console entry point.

* **Bug Fixes**
  * Improved error classification: gRPC "not found" now maps to not_found.

* **Documentation**
  * Installation and architecture docs updated with server install/usage and REST surface.

* **Tests / CI**
  * Comprehensive server tests, guardrails, in-memory fakes, and a dedicated CI job added.

### CODE REVIEW

**Comment by teng-lin** [comment]
@claude review

File: src/notebooklm/server/routes/sources.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+    The per-chunk size check below caps the copy into *our* temp file; the
+    primary disk-exhaustion guard is the Content-Length pre-check in the app
+    middleware (see ``app.py``). For a chunked (no-Content-Length) upload that
+    bypasses the pre-check, Starlette has already spooled the part before this
+    runs, so the check is a backstop on our own write, not on Starlette's spool.
+    """
+    suffix = os.path.splitext(file.filename or "")[1]
+    fd, temp_path = tempfile.mkstemp(suffix=suffix, prefix="nblm-upload-")
+    try:
+        os.fchmod(fd, 0o600)
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/server/routes/sources.py]: Addressed in 2b251839: guarded the call with `if hasattr(os, "fchmod")`. os.fchmod is Unix-only; on Windows mkstemp already creates the file owner-only. Also added a dedicated `REST server` CI job (ubuntu/macos/windows) installing the `server` extra — the canonical install omits fastapi, so `tests/server` was being skipped in CI and this Windows-only AttributeError went undetected. Good catch.

File: src/notebooklm/server/__main__.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+def _resolve_port(raw: str) -> int:
+    """Convert the (possibly env-derived) ``--port`` string to an int, or fail clean."""
+    try:
+        return int(raw)
+    except (TypeError, ValueError):
+        raise SystemExit(
+            f"Invalid port {raw!r}: must be an integer "
+            f"(check the --port argument and NOTEBOOKLM_SERVER_PORT)."
+        ) from None
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/server/__main__.py:131]: Addressed in 2b251839: `_resolve_port` now rejects ports outside 0..65535 at parse time with the same clean SystemExit path, plus tests for the out-of-range and boundary (0, 65535) cases.

File: tests/server/test_hardening.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+        headers = {"Authorization": "Bearer test-token", "Host": "127.0.0.1"}
+        with TestClient(app, headers=headers, raise_server_exceptions=False) as c:
+            resp = c.post(
+                "/v1/notebooks/nb-1/sources/file",
+                files={"file": ("big.txt", b"x" * 4096, "text/plain")},
+            )
+        assert resp.status_code == 413
+        assert called["add_file"] is False
+        assert resp.json()["error"]["category"] == "VALIDATION"
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply tests/server/test_hardening.py]: Addressed in 2b251839: the pre-buffer 413 now routes through the shared `http_error_response` projector, so it emits the lowercase `validation` category like every other error response (it was a hand-built uppercase `VALIDATION`). Test updated to assert lowercase.

File: tests/server/conftest.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
+"""Shared fixtures for the REST server test suite.
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply tests/server/conftest.py:1]: Declining this one, with reasoning. `tests/server/` is a deliberate, cohesive adapter suite (mirroring the `notebooklm.server` package the same way the CLI/MCP adapters keep their own suites), gated as a unit by `pytest.importorskip("fastapi")` so the whole directory skips cleanly when the optional `server` extra is absent. Splitting it across `tests/unit|integration|e2e/` would fragment that single importorskip guard and the shared TestClient/fake-client fixtures across three trees with no behavioral gain.

On the spirit of the taxonomy (no-network): the suite is network-free by construction — it drives a fake in-memory client via `create_app(client_factory=…)`, and `tests/_guardrails/test_server_boundary.py` enforces that `notebooklm.server` never imports `click`/`rich`/`cli`. So it satisfies the unit/integration intent (no real API) while staying co-located with the adapter it covers. The new `REST server` CI job (added in this push) runs it across ubuntu/macos/windows under the `server` extra. Happy to revisit if the team wants an explicit `tests/adapters/` convention, but that's a repo-wide decision beyond this PR.

**Comment by teng-lin** [comment]
@claude review
